### PR TITLE
Initial draft for nighthawk workload

### PR DESCRIPTION
**Pull request by vishnuchalla**
### Description
Adding nighthawk as a workload to the operator.
Wrapper side of it for more context: https://github.com/cloud-bulldozer/benchmark-wrapper/pull/446

#### Sample CR Configuration:
```
apiVersion: ripsaw.cloudbulldozer.io/v1alpha1
kind: Benchmark
metadata:
  name: nighthawk-benchmark
  namespace: benchmark-operator
spec:
  clustername: myk8scluster
  elasticsearch:
    url: https://search-perfscale-dev-chmf5l4sh66lvxbnadi4bznl3a.us-west-2.es.amazonaws.com:443
  workload:
    cleanup: true
    name: nighthawk
    args:
      # url: https://www.google.com
      image: quay.io/vchalla/nighthawk:latest
      terminations: ["http", "edge", "passthrough", "reencrypt"]
      kind: pod
      hostnetwork: false
      number_of_routes: 3
      samples: 1
      concurrency: 8
      duration: 60
      connections: 80 
      max_requests_per_connection: 50
      debug: true

```
#### CR for testing default values:
```
apiVersion: ripsaw.cloudbulldozer.io/v1alpha1
kind: Benchmark
metadata:
  name: nighthawk-benchmark
  namespace: benchmark-operator
spec:
  clustername: myk8scluster
  elasticsearch:
    url: https://search-perfscale-dev-chmf5l4sh66lvxbnadi4bznl3a.us-west-2.es.amazonaws.com:443
  workload:
    cleanup: true
    name: nighthawk
    args:
      terminations: ["http"]
```

### Testing
Tested and verified the results in the ES and pod logs as well.
#### Sample ES docs
```
"hits" : [
      {
        "_index" : "ripsaw-nighthawk-results",
        "_type" : "_doc",
        "_id" : "4a2a5aa5c41a3ec53bdc5de9f9ea4a04f12cc8fa967fa2033b95f82e2ae356f0",
        "_score" : 1.0,
        "_source" : {
          "concurrency" : 8,
          "duration" : 60,
          "connections" : 80,
          "max_requests_per_connection" : 4294937295,
          "rps" : 5,
          "kind" : "pod",
          "url" : "http://nighthawk-benchmark-route-http-2-benchmark-operator.apps.vchalla-perfscale.perfscale.devcluster.openshift.com",
          "workload" : "nighthawk",
          "uuid" : "e777d5e4-4b9e-5e90-a453-61e30caa9a5b",
          "user" : "ripsaw",
          "cluster_name" : "myk8scluster",
          "targets" : [
            "http://nighthawk-benchmark-route-http-2-benchmark-operator.apps.vchalla-perfscale.perfscale.devcluster.openshift.com"
          ],
          "hostname" : "nighthawk-client-10.0.221.20-nginx-http-2-e777d5e4-pmxvt",
          "requested_qps" : 40,
          "throughput" : 39.98936405750067,
          "status_codes_1xx" : 0,
          "status_codes_2xx" : 2400,
          "status_codes_3xx" : 0,
          "status_codes_4xx" : 0,
          "status_codes_5xx" : 0,
          "p50_latency" : 3.188863,
          "p75_latency" : 4.302335,
          "p80_latency" : 5.629951,
          "p90_latency" : 10.191871,
          "p95_latency" : 13.958143,
          "p99_latency" : 24.370175,
          "p99_9_latency" : 39.522303,
          "avg_latency" : 4.663531,
          "timestamp" : "2022-08-18T19:57:57.357909411Z",
          "bytes_in" : 1108800.0,
          "bytes_out" : 331200.0,
          "iteration" : 1,
          "run_id" : "NA"
        }
      }
    ]
```
#### Pods logs can be verified using below script once all the clients are reached to completion state
```
echo -n "`oc get all | grep -i -e 'completed' | cut -f 1 -d ' '`" > output.txt

cat output.txt | awk '{print $0}' | while
	read each
do
	echo "client name: $each"
	echo "client info"
	echo -n "`oc get $each -o yaml | grep -i -e "clientfor" -e 'port' -A 2`"
	echo -e "\n"
	echo "client logs"
	echo -n "`oc logs $each | tail -n 5`"
	echo -e "\n\n"
done
```
#### Some of the Testing Screenshots:
Screenshot from 2022-08-18 17-41-27

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -39,10 +39,10 @@ ARCH ?= $(shell uname -m | sed 's/x86_64/amd64/')
 
 # Image URL to use all building/pushing image targets
 REGISTRY ?= quay.io
-ORG ?= cloud-bulldozer
+ORG ?= vchalla
```
Comment (dry923) [inline Makefile:42]: Needs to be reverted before merging

File: roles/nighthawk/tasks/edge-tasks.yml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+---
+# tasks for edge termination type.
+
+- name: Create service for edge type server pods
+  k8s:
+    definition: "{{ lookup('template', 'http-service.yml.j2') | from_yaml }}"
+  with_sequence: start=0 count={{ workload_args.number_of_routes | int }}
```
Comment (dry923) [inline roles/nighthawk/tasks/edge-tasks.yml:7]: Multiple instances of workload_args (such as number_of_routes) being referenced without a default value.

File: roles/nighthawk/templates/workload.yml.j2 [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+               fi;
+               break;
+             done;
+             redis-cli -h {{bo.resources[0].status.podIP}} set start-{{ trunc_uuid }} false;
+             exit $exit_status;
+      restartPolicy: Never
+{% if workload_args.pin is sameas true %}
+      nodeSelector:
+          kubernetes.io/hostname: '{{ workload_args.pin_client }}'
+{% endif %}
```
Comment (dry923) [inline roles/nighthawk/templates/workload.yml.j2:135]: Should include the metadata-collection include statement like https://github.com/cloud-bulldozer/benchmark-operator/blob/0e6043a1473c895ea381712effa35e4c27808850/roles/uperf/templates/server.yml.j2#L70

File: Makefile (hunk 1 above)
Comment (vishnuchalla) [reply Makefile]: Corrected it.

File: roles/nighthawk/tasks/edge-tasks.yml (hunk 2 above)
Comment (vishnuchalla) [reply roles/nighthawk/tasks/edge-tasks.yml]: Added default values to all the workload arguments

**Review by dry923 — APPROVED** [review]
LGTM

**Comment by dry923** [comment]
@rsevilla87 @mukrishn @jtaleric PTAL when you have a chance

File: roles/nighthawk/templates/http-service.yml.j2 [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  namespace: "{{ operator_namespace }}"
+spec:
+  ports:
+  - name: http
+    port: 8080
+    protocol: TCP
+    targetPort: 8080
+  selector:
+    app: nginx-{{ type }}-{{ item }}
+  type: {{ workload_args.service_type | default("ClusterIP") }}
```
Comment (rsevilla87) [inline roles/nighthawk/templates/http-service.yml.j2:16]: You can remove all references to NodePort and just left the default, ClusterIP, ingress traffic flow is not affected by the type of service backing the route(s)

File: roles/nighthawk/README.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
+nighthawk-benchmark
```
Comment (rsevilla87) [inline roles/nighthawk/README.md:1]: Missing some docs here :)

File: roles/nighthawk/templates/http-service.yml.j2 (hunk 4 above)
Comment (vishnuchalla) [reply roles/nighthawk/templates/http-service.yml.j2]: Keeping ClusterIP as default as per the suggestion. Thanks

File: roles/nighthawk/README.md (hunk 5 above)
Comment (vishnuchalla) [reply roles/nighthawk/README.md:1]: Added docs. Thanks for pointing out.

File: roles/nighthawk/tasks/http-tasks.yml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+- name: Create service for http type server pods
+  k8s:
+    definition: "{{ lookup('template', 'http-service.yml.j2') | from_yaml }}"
+  with_sequence: start=0 count={{ workload_args.number_of_routes | default(1)  | int }}
+  vars:
+    type: http
+  when: 
+    - workload_args.serviceip is defined and workload_args.serviceip
+    - ( workload_args.service_type | default("ClusterIP") == "NodePort" )
```
Comment (rsevilla87) [inline roles/nighthawk/tasks/http-tasks.yml:12]: I think you can remove `| default("ClusterIP") == "NodePort"` ?

File: roles/nighthawk/tasks/edge-tasks.yml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+- name: Create service for edge type server pods
+  k8s:
+    definition: "{{ lookup('template', 'http-service.yml.j2') | from_yaml }}"
+  with_sequence: start=0 count={{ workload_args.number_of_routes | default(1)  | int }}
+  vars:
+    type: edge
+  when: 
+    - workload_args.serviceip is defined and workload_args.serviceip
+    - ( workload_args.service_type | default("ClusterIP") == "NodePort" )
```
Comment (rsevilla87) [inline roles/nighthawk/tasks/edge-tasks.yml:12]: I think you can remove this line?

File: roles/nighthawk/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    
+    - name: Start Client(s) with nodeport serviceIP
+      k8s:
+        definition: "{{ lookup('template', 'workload.yml.j2') | from_yaml }}"
+      with_items: "{{ server_pods.resources }}"
+      vars:
+        routerCanonicalName: "{{ routes.resources[0] | json_query('status.ingress[0].routerCanonicalHostname') }}"
+      when: 
+        - ( workload_args.serviceip|default(False) == True and server_pods.resources|length > 0 and 
+            workload_args.service_type | default("ClusterIP") == "NodePort" )
```
Comment (rsevilla87) [inline roles/nighthawk/tasks/main.yml:83]: This block can be removed, NodePort is not available anymore

File: roles/nighthawk/tasks/passthrough-tasks.yml [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+- name: Create service for passthrough type server pods
+  k8s:
+    definition: "{{ lookup('template', 'https-service.yml.j2') | from_yaml }}"
+  with_sequence: start=0 count={{ workload_args.number_of_routes | default(1)  | int }}
+  vars:
+    type: passthrough
+  when: 
+    - workload_args.serviceip is defined and workload_args.serviceip
+    - ( workload_args.service_type | default("ClusterIP") == "NodePort" )
```
Comment (rsevilla87) [inline roles/nighthawk/tasks/passthrough-tasks.yml:12]: Remove this line?

File: roles/nighthawk/tasks/reencrypt-tasks.yml [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+- name: Create service for reencrypt type server pods
+  k8s:
+    definition: "{{ lookup('template', 'https-service.yml.j2') | from_yaml }}"
+  with_sequence: start=0 count={{ workload_args.number_of_routes | default(1)  | int }}
+  vars:
+    type: reencrypt
+  when: 
+    - workload_args.serviceip is defined and workload_args.serviceip
+    - ( workload_args.service_type | default("ClusterIP") == "NodePort" )
```
Comment (rsevilla87) [inline roles/nighthawk/tasks/reencrypt-tasks.yml:12]: Remove this line?

File: roles/nighthawk/templates/workload.yml.j2 [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+kind: Job
+apiVersion: batch/v1
+metadata:
+{% if workload_args.serviceip is sameas true %}
+{% if workload_args.service_type | default("ClusterIP") == "NodePort" %}
+  name: 'nighthawk-client-{{item.status.hostIP}}-{{ item.metadata.labels.app }}-{{ trunc_uuid }}'
+{% endif %}
+{% else %}
+  name: 'nighthawk-client-{{item.status.podIP}}-{{ trunc_uuid }}'
+{% endif %}
```
Comment (rsevilla87) [inline roles/nighthawk/templates/workload.yml.j2:11]:
```suggestion
  name: 'nighthawk-client-{{item.status.podIP}}-{{ trunc_uuid }}'
```

Comment (rsevilla87) [inline roles/nighthawk/templates/workload.yml.j2:11]:
```suggestion
  name: 'nighthawk-client-{{item.status.podIP}}-{{ trunc_uuid }}'
```

Comment (vishnuchalla) [reply roles/nighthawk/templates/workload.yml.j2]: Keeping the name in the "if" clause as it more verbose and understandable while debugging.

File: roles/nighthawk/tasks/reencrypt-tasks.yml (hunk 10 above)
Comment (vishnuchalla) [reply roles/nighthawk/tasks/reencrypt-tasks.yml]: Done as we have decided to keep clusterIP as default

File: roles/nighthawk/tasks/passthrough-tasks.yml (hunk 9 above)
Comment (vishnuchalla) [reply roles/nighthawk/tasks/passthrough-tasks.yml]: Done as we have decided to keep clusterIP as default

File: roles/nighthawk/tasks/main.yml (hunk 8 above)
Comment (vishnuchalla) [reply roles/nighthawk/tasks/main.yml]: Done as we have decided to keep clusterIP as default

File: roles/nighthawk/tasks/edge-tasks.yml (hunk 7 above)
Comment (vishnuchalla) [reply roles/nighthawk/tasks/edge-tasks.yml]: Done as we have decided to keep clusterIP as default

File: roles/nighthawk/tasks/http-tasks.yml (hunk 6 above)
Comment (vishnuchalla) [reply roles/nighthawk/tasks/http-tasks.yml]: Done as we have decided to keep clusterIP as default

**Comment by vishnuchalla** [comment]
Tested again and verified results with all the new changes.

**Review by rsevilla87 — APPROVED** [review]
lgtm!

**Review by rsevilla87 — APPROVED** [review]
lgtm

**Pull request merged**